### ISSUE TITLE
mypy no longer allows int/float operations with tensors

### ISSUE THREAD

**Issue by adamjstewart** [pytorch/pytorch#116328]
### 🐛 Describe the bug

This bug started occurring with mypy 1.7+, so I'm not sure if this is a bug in PyTorch or a bug in mypy. To reproduce, create the following `test.py` file:
```python
import torch

x = torch.rand(2, 2)
reveal_type(x)
y = x + 1
reveal_type(y)
```
Then run:
```console
> mypy --strict test.py
test.py:4: note: Revealed type is "torch._tensor.Tensor"
test.py:6: note: Revealed type is "builtins.int"
```
The type of y should be Tensor as well. This is true in mypy 1.6 and older, and when not using strict mode.

### Versions

I don't use pip so the script doesn't work, but here are some relevant versions:

* Python 3.11.6
* PyTorch 2.1.2
* mypy 1.8.0

cc @ezyang @malfet @xuzhao9 @gramster

**Comment by adamjstewart** [comment]
This also occurs for subtraction, multiplication, and division. Also occurs for == (which should return a boolean Tensor, but mypy thinks it returns a bool)

**Comment by tringwald** [comment]
I can reproduce this on `torch==2.1.2`, but not anymore on latest `main` builds (with neither `mypy==1.7` nor `1.8`).

```bash
$ mypy --strict test.py 
test.py:4:13: note: Revealed type is "torch._tensor.Tensor"
test.py:6:13: note: Revealed type is "torch._tensor.Tensor"
Success: no issues found in 1 source file
```

**Comment by adamjstewart** [comment]
Thanks, maybe it was already fixed? It looks like 2.2 is already in the works. Will confirm once it's released.

**Comment by adamjstewart** [comment]
It appears we're running mypy 1.7 in CI, so this should definitely be fixed on main. I think we can close. Thanks for the help!

**Comment by adamjstewart** [comment]
Can confirm that this issue was fixed in PyTorch 2.2!
